common/persistence/client/health_request_rate_limiter.go

```diff
@@ -0,0 +1,132 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/quotas"
+)
+
+const (
+	DefaultRateBurstRatio    = 1.0
+	DefaultMinRateMultiplier = 0.1
+	DefaultMaxRateMultiplier = 1.0
+)
+
+type (
+	HealthRequestRateLimiterImpl struct {
+		rateLimiter   *quotas.RateLimiterImpl
+		healthSignals persistence.HealthSignalAggregator
+
+		refreshTimer *time.Ticker
+
+		// thresholds which should trigger backoff if exceeded
+		latencyThreshold float64
+		errorThreshold   float64
+
+		rateFn           quotas.RateFn
+		rateToBurstRatio float64
+
+		minRateMultiplier float64
+		maxRateMultiplier float64
+		curRateMultiplier float64
+
+		// if either threshold is exceeded, the current rate multiplier will be reduced by this amount
+		rateBackoffStepSize float64
+		// when the system is healthy and current rate < max rate, the current rate multiplier will be
+		// increased by this amount
+		rateIncreaseStepSize float64
+	}
+)
+
+var _ quotas.RequestRateLimiter = (*HealthRequestRateLimiterImpl)(nil)
+
+func NewHealthRequestRateLimiterImpl(
+	healthSignals persistence.HealthSignalAggregator,
+	refreshInterval time.Duration,
+	rateFn quotas.RateFn,
+	latencyThreshold float64,
+	errorThreshold float64,
+	rateBackoffStepSize float64,
+	rateIncreaseStepSize float64,
+) *HealthRequestRateLimiterImpl {
+	return &HealthRequestRateLimiterImpl{
+		rateLimiter:          quotas.NewRateLimiter(rateFn(), int(DefaultRateBurstRatio*rateFn())),
+		healthSignals:        healthSignals,
+		refreshTimer:         time.NewTicker(refreshInterval),
+		rateFn:               rateFn,
+		rateToBurstRatio:     DefaultRateBurstRatio,
+		latencyThreshold:     latencyThreshold,
+		errorThreshold:       errorThreshold,
+		minRateMultiplier:    DefaultMinRateMultiplier,
+		maxRateMultiplier:    DefaultMaxRateMultiplier,
+		curRateMultiplier:    DefaultMaxRateMultiplier,
+		rateBackoffStepSize:  rateBackoffStepSize,
+		rateIncreaseStepSize: rateIncreaseStepSize,
+	}
+}
+
+func (rl *HealthRequestRateLimiterImpl) Allow(now time.Time, request quotas.Request) bool {
+	rl.maybeRefresh()
+	return rl.rateLimiter.AllowN(now, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) Reserve(now time.Time, request quotas.Request) quotas.Reservation {
+	rl.maybeRefresh()
+	return rl.rateLimiter.ReserveN(now, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) Wait(ctx context.Context, request quotas.Request) error {
+	rl.maybeRefresh()
+	return rl.rateLimiter.WaitN(ctx, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) maybeRefresh() {
+	select {
+	case <-rl.refreshTimer.C:
```

Comment:
also add a channel from persistence.HealthSignalAggregator to rate limiter for quicker reaction?

Reply:
I think we do not necessarily need the fastest reaction. I want to avoid the situation where the rate limit seesaws up and down too frequently.